File: shopee/src/main/kotlin/kuwelym/shopee/domain/entities/Product.kt
package kuwelym.shopee.domain.entities

import java.math.BigDecimal
import java.time.LocalDateTime

data class Product(
    val id: String? = null,
    val name: String,
    val description: String,
    val price: BigDecimal,
    val stock: Int,
    val shopId: String,
    val categoryId: String? = null,
    val imageUrls: MutableList<String> = mutableListOf(),
    val createdAt: LocalDateTime = LocalDateTime.now(),
    val updatedAt: LocalDateTime = LocalDateTime.now(),
    val isActive: Boolean = true,
) {
    fun isValidPrice(): Boolean = price > BigDecimal.ZERO

    fun isInStock(): Boolean = stock > 0

    fun hasEnoughStock(quantity: Int): Boolean = stock >= quantity

    fun reduceStock(quantity: Int): Product {
        require(hasEnoughStock(quantity)) { "Insufficient stock" }
        return this.copy(stock = stock - quantity)
    }

    fun addStock(quantity: Int): Product {
        require(quantity > 0) { "Quantity must be positive" }
        return this.copy(stock = stock + quantity)
    }

    fun updatePrice(newPrice: BigDecimal): Product {
        require(newPrice > BigDecimal.ZERO) { "Price must be positive" }
        return this.copy(price = newPrice, updatedAt = LocalDateTime.now())
    }
}

File: shopee-front-end/src/app/types/product.ts
export interface Product {
  id?: string;
  name: string;
  description: string;
  price: number;
  stock: number;
  shopId: string;
  categoryId?: string;
  imageUrls: string[];
  createdAt: string;
  updatedAt: string;
  isActive: boolean;
}

export interface CreateProductRequest {
  name: string;
  description: string;
  price: number;
  stock: number;
  categoryId?: string;
  imageUrls: string[];
}

export interface UpdateProductRequest {
  name?: string;
  description?: string;
  price?: number;
  stock?: number;
  categoryId?: string;
  imageUrls?: string[];
}

export type ProductResponse = Product;

export interface ProductsApiResponse {
  success: boolean;
  message: string;
  data?: Product[];
}

export interface ProductApiResponse {
  success: boolean;
  message: string;
  data?: Product;
} 
